text/0079-exception-groups.md

```diff
@@ -0,0 +1,348 @@
+on an event. One or more exceptions can be sent in the `values` array of the interface.  Multiple values represent
+a chain of exceptions in a causal relationship, sorted from oldest to newest.  For example:
+
+```json
+{
+  "exception": {
+    "values": [
+      {"type": "TypeError", "value": "Invalid Type!"},
+      {"type": "ValueError", "value": "Invalid Value!"},
+      {"type": "RuntimeError", "value": "Something went wrong!"}
+    ]
+  }
+}
+```
+
+In the above example, an issue is created in Sentry for the `RuntimeError` exception.
+
+```
+Issue Title: "RuntimeError: Something went wrong!"
+
++==============================
++ RuntimeError: Something went wrong!
++------------------------------
++ (stack trace)
++==============================
+
++==============================
++ ValueError: Invalid Value!
++------------------------------
++ (stack trace)
++==============================
+
++==============================
++ TypeError: Invalid Type!
++------------------------------
++ (stack trace)
++==============================
+```
+
+This design is linear, and thus cannot support exception groups having more than one exception.
+But even in that case, the exception presented in Sentry would not be titled or grouped correctly.
+
+Consider:
+
+```json
+{
+  "exception": {
+    "values": [
+      {"type": "TypeError", "value": "Invalid Type!"},
+      {"type": "ValueError", "value": "Invalid Value!"},
+      {"type": "RuntimeError", "value": "Something went wrong!"},
+      {"type": "exception_group", "value": "Exception Group (1 sub-exception)"}
+    ]
+  }
+}
+```
+
+Sending this to Sentry would result in an issue titled `"exception_group: Exception Group (1 sub-exception)"`,
+which is unactionable.
+
+# Supporting Data
+
+- [Python customer feedback](https://github.com/getsentry/sentry/issues/37716)
+- [.NET customer feedback](https://github.com/getsentry/sentry-dotnet/issues/270)
+
+# Options Considered
+
+## Option 1
+
+Do nothing.
+
+Pros:
+- Nothing to do.
+
+Cons:
+- Issues are titled and grouped by the exception group instead of an actionable exception.
+- Exception groups with the same sub-exceptions but different quantities of them will be grouped into separate issues.
+- Exception groups with different sub-exceptions will be grouped into separate issues with identical titles.
+
+## Option 2
+
+Without any modification to ingest or UI, the SDKs can attempt to unwind exception groups,
+flattening them into a single list of exceptions.  Then, the containing exception group can
+(optionally) be discarded.
+
+The .NET SDK currently does this (as of v3.18.0, June 2022), but it is considered a hack and
+has many disadvantages.
+
+Pros:
+- SDK changes are straightforward.
+- Already implemented in the .NET SDK.
+- Doesn't require any protocol, ingest, or UI changes.
+
+Cons:
+- Doesn't accurately represent the exception group.
+- Implies a causal relationship between sibling exception that doesn't exist.
```

Comment:
How would you expect the error event to be grouped issues for Options 2 and 4?

Reply:
Option 2 is what .NET does today (and shouldn't, IMHO).  I don't really want to pursue this, but included it for completeness.  By default it takes one exception out of the group and only captures that.  All others are flattened below it (incorrectly).  So the issue is titled based on that exception - but it's grouped on the entire set of exceptions.  (Oops - one of the cons on grouping I accidentally put under option 1 instead of 2. Will fix.)

Also, .NET provides `SentryOptions.KeepAggregateException` which is `false` by default, but when set `true` will include the exception group.  So in that case, the title of the issue is the meaningless message about the exception group, and that is also used in issue grouping.

Both are bad.

Reply:
For Option 4, grouping would be similar to Option 3, but would have to take nested exception groups into account somehow.  I'm really not sure how, TBH.  Maybe a strong reason not to do Option 4.

Reply:
The RFC has been changed considerably.  The revised proposal does not require any changes to how grouping works.  Thanks.